src/OrleansProviders/Streams/Memory/MemoryBatchContainer.cs

```diff
@@ -60,5 +57,10 @@ public bool ShouldDeliver(IStreamIdentity stream, object filterData, StreamFilte
         {
             return true;
         }
+
+        void IOnDeserialized.OnDeserialized(ISerializerContext context)
+        {
+            this.serializer = ActivatorUtilities.GetServiceOrCreateInstance<TSerializer>(context.ServiceProvider);
```

Comment:
Awesome, this is what I was advocating for. Nevertheless, for performance reasons, I think there is a method in ActivatorUtilities that allows you to create an object factory and cache it, and avoid reflection in subsequent calls, since this is going to be called extremely frequently for these same types